### ISSUE TITLE
[dask] Support custom metric functions

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#3956]
## Summary

The Dask estimators in `lightgbm.dask` should support the use of a custom objective function.

## Motivation

Defining custom metrics functions gives users fine-grained ability to evaluate what's happening during training, and to control the training process via early stopping (#3712).

This feature would also bring Dask estimators closer to parity with the sklearn estimators.

## Description

I haven't thought through this much yet, just writing up a placeholder issue for discussion. If you're reading this and have ideas, please comment and the issue can be re-opened.

## References

See LINK for an explanation of how this works in the sklearn estimators.

Can look at how xgboost.dask handles this for inspiration:

* https://github.com/dmlc/xgboost/blob/4c5d2608e0bd179bfc5a8e9b3132e6651f841256/python-package/xgboost/dask.py#L1565
* https://github.com/dmlc/xgboost/blob/4c5d2608e0bd179bfc5a8e9b3132e6651f841256/python-package/xgboost/dask.py#L1838-L1841
* https://github.com/dmlc/xgboost/blob/4c5d2608e0bd179bfc5a8e9b3132e6651f841256/python-package/xgboost/dask.py#L850

At a quick glance it seems like `xgboost.dask` just passes the function all the way through to training and it "just works". Maybe we'll get lucky 😂

**Comment by jameslamb** [comment]
Closing this in favor of putting it in #2302 with other feature requests. Please comment if you're interested in working on this and the issue can be re-opened.

**Comment by jameslamb** [comment]
Re-opening this. I think it can be closed by a pull request that is currently in progress: https://github.com/microsoft/LightGBM/pull/4101#discussion_r603754561

**Comment by jameslamb** [comment]
fixed by #4101, thanks @ffineis !
